File: Google/Golang/binary-tree-postorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func postorderTraversal(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    stack := []*TreeNode{root}
    output := make([]int, 0)
    for true {
        for root != nil {
            if root.Right != nil {
                stack = append(stack, root.Right)
            }
            stack = append(stack, root)
            root = root.Left
        }
        root = stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        if root.Right != nil && len(stack) > 0 && root.Right == stack[len(stack)-1] {
            stack = stack[:len(stack)-1]
            stack = append(stack, root)
            root = root.Right
        } else if len(stack) > 0 {
            output = append(output, root.Val)
            root = nil
        }
        if len(stack) == 0 {
            break
        }
    }
    return output
}

File: Tree/C++/binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        TreeNode *node;
        stack<TreeNode*> s;
        if (root == NULL) {
            return res;
        }
        s.push(root);
        while (!s.empty()) {
            node = s.top();
            s.pop();
            res.push_back(node->val);
            if (node->left) {
                s.push(node->left);
            }
            if (node->right) {
                s.push(node->right);
            }
        }
        reverse(res.begin(), res.end());
        return res;
    }
};
